bin/manage.py

```diff
@@ -0,0 +1,406 @@
+            # to tag each stack with an identifier for that process
+            arg = "[{}]".format(sys.argv[1])
+        except IndexError:
+            arg = "[pre_start]"
+        name = '{}{}{}'.format(arg, (len(inspect.stack()) * " "), fn.__name__)
+        log.debug('%s' % name)
+        out = apply(fn, args, kwargs)
+        log.debug('%s: %s', name, out)
+        return out
+    return wrapper
+
+def get_environ(key, default):
+    """
+    Gets an environment variable and trims away comments and whitespace.
+    """
+    val = os.environ.get(key, default)
+    try:
+        val = val.split('#')[0]
+        val = val.strip()
+    finally:
+        # just swallow AttributeErrors for non-strings
+        return val
+
+# ---------------------------------------------------------
+
+SESSION_CACHE_FILE = get_environ('SESSION_CACHE_FILE', '/tmp/mongodb-session')
+SESSION_NAME = get_environ('SESSION_NAME', 'mongodb-replica-set-lock')
+SESSION_TTL = int(get_environ('SESSION_TTL', 60))
+
+# consts for node state
+PRIMARY = 'mongodb-primary'
+SECONDARY = 'mongodb-secondary'
+REPLICA = 'mongodb'
+
+# key where primary will be stored in consul
+PRIMARY_KEY = get_environ('PRIMARY_KEY', 'mongodb-primary')
+
+# how many time to retry connecting to mongo
+# be aware that the health check may timeout before this is reached
+MONGO_RETRY_TIMES=int(get_environ('MONGO_RETRY_TIMES', 10))
+
+# ---------------------------------------------------------
+
+class ContainerPilot(object):
+    """
+    ContainerPilot config is where we rewrite ContainerPilot's own config
+    so that we can dynamically alter what service we advertise
+    """
+
+    def __init__(self):
+        # TODO: we should make sure we can support JSON-in-env-var
+        # the same as ContainerPilot itself
+        self.path = get_environ('CONTAINERPILOT', None).replace('file://', '')
+        with open(self.path, 'r') as f:
+            self.config = json.loads(f.read())
+
+    @debug
+    def update(self, state):
+        if state and self.config['services'][0]['name'] != state:
+            self.config['services'][0]['name'] = state
+            self.render()
+            return True
+
+    @debug
+    def render(self):
+        new_config = json.dumps(self.config)
+        with open(self.path, 'w') as f:
+            f.write(new_config)
+
+    def reload(self):
+        """ force ContainerPilot to reload its configuration """
+        log.info('Reloading ContainerPilot configuration.')
+        os.kill(1, signal.SIGHUP)
+
+# ---------------------------------------------------------
+# Top-level functions called by ContainerPilot or forked by this program
+
+@debug
+def pre_start():
+    """
+    MongoDB must be running in order to execute most of our setup behavior
+    so we're just going to make sure the directory structures are in
+    place and then let the first health check handler take it from there
+    """
+    # TODO is there anything that needs to be done before starting mongo?
```

Comment:
@tianon and I had a conversation about abstracting the object storage so that we could better support local development and pluggability of the object store. https://github.com/autopilotpattern/mysql/issues/35 covers that vaguely for autopilotpattern/mysql.

@yosifkit have you had any thoughts on how to implement such a thing?